### PR TITLE
Validate Network for ProposalProcedure and TreasuryWithdrawal

### LINKED ISSUE #3603: Validate NetworkIds
Ensure that all RewardAcnt that were added in ConwayEra have their NetworkIds verified:

* [x] - ProposalProcedure
* [x] - TreasuryWithdrawal

### PR DESCRIPTION
**Pull request by aniketd**
# Description

Resolves #3603 

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Enact.hs [hunk 1]
```diff
@@ -50,19 +52,20 @@ instance EraGov era => STS (ConwayENACT era) where
...
-          -- TODO: validate NetworkId in all
-          -- We can drop NetworkIds after their validation
-          wdrlsNoNetworkId = Map.mapKeys getRwdCred wdrls
+      expectedNetworkId <- liftSTS $ asks networkId
+      let withdrawals = Map.filterWithKey (\rwrdAcnt _ -> getRwdNetwork rwrdAcnt == expectedNetworkId) wdrls
+          wdrlsAmount = fold withdrawals
+          wdrlsNoNetworkId = Map.mapKeys getRwdCred withdrawals
       runTest
         . failureUnless (wdrlsAmount <= ensTreasury st)
         $ EnactTreasuryInsufficientFunds @era wdrls (ensTreasury st)
```
Comment (aniketd) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Enact.hs]: `wdrls` here is intentionally left unedited. Should this instead be just the filtered ones?

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Enact.hs [hunk 2]
```diff
@@ -50,19 +52,20 @@ instance EraGov era => STS (ConwayENACT era) where
...
     HardForkInitiation _prevGovActionId pv -> pure $ st {ensProtVer = pv}
     TreasuryWithdrawals wdrls -> do
-      let wdrlsAmount = fold wdrls
-          -- TODO: validate NetworkId in all
-          -- We can drop NetworkIds after their validation
-          wdrlsNoNetworkId = Map.mapKeys getRwdCred wdrls
+      expectedNetworkId <- liftSTS $ asks networkId
+      let withdrawals = Map.filterWithKey (\rwrdAcnt _ -> getRwdNetwork rwrdAcnt == expectedNetworkId) wdrls
+          wdrlsAmount = fold withdrawals
+          wdrlsNoNetworkId = Map.mapKeys getRwdCred withdrawals
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Enact.hs]: NetworkId for the withdrawal account need to be validated upon proposal, i.e. in the `GOV` rule, not during Enactment.

If it is validated upon proposal, then here, when we ENACT this withdrawal action we can simply ignore the NetworkId value

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs [hunk 3]
```diff
@@ -63,6 +66,7 @@ data GovEnv era = GovEnv
 data ConwayGovPredFailure era
   = GovActionsDoNotExist (Set.Set (GovActionId (EraCrypto era)))
   | MalformedProposal (GovAction era)
+  | WrongNetworkIdInReturnAddressOfProposalProcedure !Network !Network -- First is supplied and second is the expected.
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs]: This is way tooooooo long. :smile: How about?
```suggestion
  | ProposalProcedureNetworkIdMismatch 
    -- | Supplied 
    !Network
    -- | Expected
    !Network
```

**Review by lehins — CHANGES_REQUESTED** [review]

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Enact.hs (hunk 2 above)
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Enact.hs]: Got it! Thanks for the pointers! :memo:

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs [hunk 4]
```diff
@@ -170,9 +177,15 @@ govTransition = do
   -- TODO Check the signatures
```
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs]: I don't think anyone knows what this TODO refers to.. maybe we can remove it?

Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs]: @Soupstraw any ideas?

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs [hunk 5]
```diff
@@ -165,14 +177,32 @@ actionWellFormed ga = failureUnless isWellFormed $ MalformedProposal ga
...
+        pProcNetworkId
+          == expectedNetworkId
+            ?! ProposalProcedureNetworkIdMismatch pProcNetworkId expectedNetworkId
+        case pProcGovAction of
+          TreasuryWithdrawals wdrls ->
+            forM_ (Map.keys wdrls) $ \rwrdAcnt ->
+              let wdrlNetworkId = getRwdNetwork rwrdAcnt
+               in wdrlNetworkId
+                    == expectedNetworkId
+                      ?! TreasuryWithdrawalsNetworkIdMismatch wdrlNetworkId expectedNetworkId
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs]: We don't need to report a failure for every account, it is better to report a set of accounts that violates this predicate. Also it is better to keep predicate check lazy. This way there is no cost payed upon replay of chain without validation.
```suggestion
            let mismatchedNetworkIds = 
                  Set.filter ((/= expectedNetworkId) . getRwdNetwork) $ Map.keys wdrls
               in Set.null mismatchedNetworkIds
                      ?! TreasuryWithdrawalsNetworkIdMismatch mismatchedNetworkIds expectedNetworkId
```

**Review by lehins — CHANGES_REQUESTED** [review]

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs (hunk 5 above)
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs]: :100: Got it!

Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs]: I didn't notice the suggestion you made here and ended up unifying the `ProposalProcedure` and `TreasuryWithdrawals` failures together. My bad. I will revert that change and apply this suggestion, before re-requesting your feedback :disappointed:

Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs]: In my high-contrast GitHub UI both the suggestions and context-snippets for comments are the same color. I should have payed more attention.

Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs]: I also misread your comment to mean that we need to report `Credential`s for which the `Network` is wrong.

Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs]: :100: Now I got it.

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs (hunk 4 above)
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs]: I will make sure to remove the comment before merging if no one yells by tomorrow. :raised_hands:

Comment (Soupstraw) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs]: Yeah I think it's not relevant

Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs]: Thansk @Soupstraw ! :raised_hands:

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Enact.hs [hunk 6]
```diff
@@ -50,18 +50,17 @@ instance EraGov era => STS (ConwayENACT era) where
   initialRules = []
   transitionRules = [enactmentTransition]
 
-enactmentTransition :: forall era. EraPParams era => TransitionRule (ConwayENACT era)
+enactmentTransition :: forall era. EraGov era => TransitionRule (ConwayENACT era)
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Enact.hs]: Reverting back to the original constraint
```suggestion
enactmentTransition :: forall era. EraPParams era => TransitionRule (ConwayENACT era)
```

**Review by lehins — APPROVED** [review]
Perfect! Thank you

**Pull request merged**